src/aviary/functional.py

```diff
@@ -0,0 +1,173 @@
+from collections.abc import Callable
+from functools import update_wrapper
+from typing import Any
+
+from pydantic import BaseModel, Field
+
+from aviary.env import Environment, Frame
+from aviary.message import Message
+from aviary.tools import Messages, Tool, ToolRequestMessage
+from aviary.utils import is_coroutine_callable
+
+
+class DynamicState(BaseModel):
+    """Dynamic env state model that adapts to provided extras."""
+
+    reward: float = 0
+    done: bool = False
+    extras: dict[str, Any] = Field(default_factory=dict)
+
+    def __getattr__(self, name: str):
+        """Allow direct access to extras as attributes."""
+        if name in self.extras:
+            return self.extras[name]
+        raise AttributeError(f"'State' has no attribute {name!r}")
+
+
+class FunctionalEnvironment(Environment[DynamicState]):
+    """Environment class for function-based environments."""
+
+    def __init__(
+        self,
+        start_fn: Callable,
+        tools: list[Tool],
+        allow_concurrency: bool,
+        *args,
+        **kwargs,
+    ):
+        super().__init__()
+        self.start_fn = start_fn
+        self.tools = tools
+        self.args = args
+        self.kwargs = kwargs
+        self.state = DynamicState()
+        self.allow_concurrency = allow_concurrency
+
+    async def _call_start(self) -> tuple[str, dict[str, Any]]:
+        """Call start_fn handling both sync and async cases."""
+        if is_coroutine_callable(self.start_fn):
+            result = await self.start_fn(*self.args, **self.kwargs)
+        else:
+            result = self.start_fn(*self.args, **self.kwargs)
+
+        if not isinstance(result, tuple) or len(result) != 2:  # noqa: PLR2004
+            raise TypeError("Start function must return (observation, state_dict)")
+
+        obs, state_dict = result
+        if not isinstance(state_dict, dict):
+            raise TypeError("State must be a dictionary")
+
+        return obs, state_dict
+
+    async def reset(self) -> tuple[Messages, list[Tool]]:
+        obs, extras = await self._call_start()
+        self.state = DynamicState(extras=extras)
+        return [Message(content=obs)], self.tools
+
+    async def step(
+        self, action: ToolRequestMessage
+    ) -> tuple[Messages, float, bool, bool]:
+        msgs = await self.exec_tool_calls(
+            action, state=self.state, concurrency=self.allow_concurrency
+        )
+        return msgs, self.state.reward, self.state.done, False
+
```

Comment:
Not a blocker at all, but could we have some basic `from_task` here? Maybe by relying on function defaults in the wrapped `self.start_fn`? 

That way these would be deployable by default.

Reply:
Per discussion - we can simply import the `EnvironmentBuilder` and use its `__call__` instead.